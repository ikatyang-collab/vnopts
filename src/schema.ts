import { VALUE_UNCHANGED } from './constants';
import {
  DefaultResult,
  DeprecatedResult,
  ExpectedResult,
  ForwardResult,
  OptionValue,
  RedirectResult,
  Utils,
  ValidateResult,
} from './types';

export interface SchemaParameters<$Value> extends SchemaHandlers<$Value> {
  name: string;
}

export interface SchemaHandlers<$Value> {
  default?:
    | DefaultResult<$Value>
    | ((schema: Schema<$Value>, utils: Utils) => DefaultResult<$Value>);
  expected?:
    | ExpectedResult
    | ((schema: Schema<$Value>, utils: Utils) => ExpectedResult);
  validate?:
    | ValidateResult
    | ((
        value: unknown,
        schema: Schema<$Value>,
        utils: Utils,
      ) => ValidateResult);
  deprecated?:
    | DeprecatedResult<$Value>
    | ((
        value: $Value,
        schema: Schema<$Value>,
        utils: Utils,
      ) => DeprecatedResult<$Value>);
  forward?:
    | ForwardResult<$Value>
    | ((
        value: $Value,
        schema: Schema<$Value>,
        utils: Utils,
      ) => ForwardResult<$Value>);
  redirect?:
    | RedirectResult<$Value>
    | ((
        value: $Value,
        schema: Schema<$Value>,
        utils: Utils,
      ) => RedirectResult<$Value>);
  overlap?: (
    currentValue: $Value,
    newValue: $Value,
    schema: Schema<$Value>,
    utils: Utils,
  ) => $Value;
  preprocess?: (
    value: unknown,
    schema: Schema<$Value>,
    utils: Utils,
  ) => unknown;
  postprocess?: (
    value: $Value,
    schema: Schema<$Value>,
    utils: Utils,
  ) => unknown;
}

const HANDLER_KEYS: Array<keyof SchemaHandlers<any>> = [
  'default',
  'expected',
  'validate',
  'deprecated',
  'forward',
  'redirect',
  'overlap',
  'preprocess',
  'postprocess',
];

export function createSchema<
  $Schema extends Schema<any>,
  $Parameters extends $Schema['_parametersType']
>(
  SchemaConstructor: new (parameters: $Parameters) => $Schema,
  parameters: $Parameters,
) {
  const schema = new SchemaConstructor(parameters);
  const subSchema = Object.create(schema) as $Schema;

  for (const handlerKey of HANDLER_KEYS) {
    if (handlerKey in parameters) {
      subSchema[handlerKey] = normalizeHandler(
        parameters[handlerKey],
        schema,
        Schema.prototype[handlerKey].length,
      );
    }
  }

  return subSchema;
}

export abstract class Schema<
  $Value extends OptionValue,
  $Parameters extends SchemaHandlers<$Value> = SchemaHandlers<$Value>
> {
  public static create<$Schema extends Schema<any>>(
    parameters: $Schema['_parametersType'],
  ): $Schema {
    // @ts-ignore: https://github.com/Microsoft/TypeScript/issues/5863
    return createSchema(this, parameters);
  }

  public name: string;

  public _valueType!: $Value;
  public _parametersType!: $Parameters;

  constructor(parameters: SchemaParameters<$Value>) {
    this.name = parameters.name;
  }

  public default(_utils: Utils): DefaultResult<$Value> {
    return undefined;
  }

  // istanbul ignore next: this is actually an abstract method but we need a placeholder to get `function.length`
  public expected(_utils: Utils): ExpectedResult {
    return 'nothing';
  }

  // istanbul ignore next: this is actually an abstract method but we need a placeholder to get `function.length`
  public validate(_value: unknown, _utils: Utils): ValidateResult {
    return false;
  }

  public deprecated(_value: $Value, _utils: Utils): DeprecatedResult<$Value> {
    return false;
  }

  public forward(_value: $Value, _utils: Utils): ForwardResult<$Value> {
    return undefined;
  }

  public redirect(_value: $Value, _utils: Utils): RedirectResult<$Value> {
    return undefined;
  }

  public overlap(
    currentValue: $Value,
    _newValue: $Value,
    _utils: Utils,
  ): $Value {
    return currentValue;
  }

  public preprocess(value: unknown, _utils: Utils): any {
    return value;
  }

  public postprocess(_value: $Value, _utils: Utils): any {
    return VALUE_UNCHANGED;
  }
}

function normalizeHandler<$Result>(
  handler: $Result | ((...args: any[]) => $Result),
  superSchema: Schema<any>,
  handlerArgumentsLength: number,
) {
  return typeof handler === 'function'
    ? (...args: any[]) =>
        handler(
          ...args.slice(0, handlerArgumentsLength - 1),
          superSchema,
          ...args.slice(handlerArgumentsLength - 1),
        )
    : () => handler;
}
