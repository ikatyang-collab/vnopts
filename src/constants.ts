export const VALUE_NOT_EXIST = Symbol.for('vnopts.VALUE_NOT_EXIST');
export const VALUE_UNCHANGED = Symbol.for('vnopts.VALUE_UNCHANGED');
