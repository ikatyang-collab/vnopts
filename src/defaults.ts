import { apiDescriptor } from './descriptors/api.js'
import { commonDeprecatedHandler } from './handlers/deprecated/common.js'
import { commonInvalidHandler } from './handlers/invalid/index.js'
import { levenUnknownHandler } from './handlers/unknown/leven.js'
import {
  DeprecatedHandler,
  Descriptor,
  InvalidHandler,
  UnknownHandler,
} from './types.js'

export const defaultDescriptor: Descriptor = apiDescriptor
export const defaultUnknownHandler: UnknownHandler = levenUnknownHandler
export const defaultInvalidHandler: InvalidHandler = commonInvalidHandler
export const defaultDeprecatedHandler: DeprecatedHandler =
  commonDeprecatedHandler
