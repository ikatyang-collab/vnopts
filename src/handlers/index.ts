export * from './deprecated/index.js'
export * from './invalid/index.js'
export * from './unknown/index.js'
