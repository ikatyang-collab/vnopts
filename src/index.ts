export * from './descriptors';
export * from './handlers';
export * from './schemas';

export * from './constants';
export * from './normalize';
export * from './schema';
export * from './types';
