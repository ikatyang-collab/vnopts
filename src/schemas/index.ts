export * from './alias.js'
export * from './any.js'
export * from './array.js'
export * from './boolean.js'
export * from './choice.js'
export * from './integer.js'
export * from './number.js'
export * from './string.js'
