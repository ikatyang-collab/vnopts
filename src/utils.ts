import {
  DefaultResult,
  DeprecatedResult,
  ForwardResult,
  InvalidHandler,
  NormalizedDefaultResult,
  NormalizedDeprecatedResult,
  NormalizedDeprecatedResultWithTrue,
  NormalizedForwardResult,
  NormalizedInvalidHandler,
  NormalizedRedirectResult,
  NormalizedTransferResult,
  NormalizedValidateResult,
  Options,
  RedirectResult,
  TransferResult,
  ValidateResult,
} from './types';

export type NotEmptyArray<T> = [T, ...T[]];

export type RecordFromUnion<T extends Options, K extends keyof T> = {
  [X in T[K]]: Extract<T, Record<K, X>>
};

export function recordFromArray<
  T extends Record<string, any>,
  K extends keyof T
>(array: T[], mainKey: K): RecordFromUnion<T, K> {
  const record = Object.create(null) as Partial<RecordFromUnion<T, K>>;

  for (const value of array) {
    const key = value[mainKey];

    // istanbul ignore next
    if (record[key]) {
      throw new Error(`Duplicate ${mainKey} ${JSON.stringify(key)}`);
    }

    // @ts-ignore
    record[key] = value;
  }

  return record as RecordFromUnion<T, K>;
}

export function mapFromArray<T extends Record<string, any>, K extends keyof T>(
  array: T[],
  mainKey: K,
): Map<T[K], T> {
  const map = new Map<T[K], T>();

  for (const value of array) {
    const key = value[mainKey];

    // istanbul ignore next
    if (map.has(key)) {
      throw new Error(`Duplicate ${mainKey} ${JSON.stringify(key)}`);
    }

    map.set(key, value);
  }

  return map;
}

export function createAutoChecklist() {
  const map: { [idString: string]: boolean } = Object.create(null);
  return (id: any) => {
    const idString = JSON.stringify(id);
    if (map[idString]) {
      return true;
    }
    map[idString] = true;
    return false;
  };
}

export function partition<T>(
  array: T[],
  predicate: (x: T) => boolean,
): [T[], T[]] {
  const trueArray: T[] = [];
  const falseArray: T[] = [];

  for (const value of array) {
    if (predicate(value)) {
      trueArray.push(value);
    } else {
      falseArray.push(value);
    }
  }

  return [trueArray, falseArray];
}

export function isInt(value: number) {
  return value === Math.floor(value);
}

export function comparePrimitive(
  a: undefined | null | boolean | number | string,
  b: undefined | null | boolean | number | string,
): number {
  if (a === b) {
    return 0;
  }

  const typeofA = typeof a;
  const typeofB = typeof b;

  const orders = [
    'undefined',
    'object', // null
    'boolean',
    'number',
    'string',
  ];
  if (typeofA !== typeofB) {
    return orders.indexOf(typeofA) - orders.indexOf(typeofB);
  }

  if (typeofA !== 'string') {
    return Number(a) - Number(b);
  }

  return (a as string).localeCompare(b as string);
}

export function normalizeInvalidHandler(
  invalidHandler: InvalidHandler,
): NormalizedInvalidHandler {
  return (...args) => {
    const errorMessageOrError = invalidHandler(...args);
    return typeof errorMessageOrError === 'string'
      ? new Error(errorMessageOrError)
      : /* istanbul ignore next*/ errorMessageOrError;
  };
}

export function normalizeDefaultResult<$Value>(
  result: DefaultResult<$Value>,
): NormalizedDefaultResult<$Value> {
  return result === undefined ? {} : result;
}

export function normalizeValidateResult(
  result: ValidateResult,
  value: unknown,
): NormalizedValidateResult {
  return result === true ? true : result === false ? { value } : result;
}

export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
): NormalizedDeprecatedResult<$Value>;
export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
  doNotNormalizeTrue: true,
): NormalizedDeprecatedResultWithTrue<$Value>;
export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
  doNotNormalizeTrue = false,
):
  | NormalizedDeprecatedResult<$Value>
  | NormalizedDeprecatedResultWithTrue<$Value> {
  return result === false
    ? false
    : result === true
      ? doNotNormalizeTrue
        ? true
        : [{ value }]
      : 'value' in result
        ? [result]
        : result.length === 0
          ? false
          : (result as NotEmptyArray<{ value: $Value }>);
}

export function normalizeTransferResult<$Value>(
  result: TransferResult<$Value>,
  value: $Value,
): NormalizedTransferResult<$Value> {
  return typeof result === 'string' || 'key' in result
    ? { from: value, to: result }
    : 'from' in result
      ? { from: result.from!, to: result.to }
      : { from: value, to: result.to };
}

export function normalizeForwardResult<$Value>(
  result: ForwardResult<$Value>,
  value: $Value,
): NormalizedForwardResult<$Value> {
  return result === undefined
    ? []
    : Array.isArray(result)
      ? result.map(transferResult =>
          normalizeTransferResult(transferResult, value),
        )
      : [normalizeTransferResult(result, value)];
}

export function normalizeRedirectResult<$Value>(
  result: RedirectResult<$Value>,
  value: $Value,
): NormalizedRedirectResult<$Value> {
  const redirect = normalizeForwardResult(
    typeof result === 'object' && 'redirect' in result
      ? result.redirect
      : result,
    value,
  );
  return redirect.length === 0
    ? { remain: value, redirect }
    : typeof result === 'object' && 'remain' in result
      ? { remain: result.remain, redirect }
      : { redirect };
}
