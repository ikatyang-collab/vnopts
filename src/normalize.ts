import { VALUE_NOT_EXIST, VALUE_UNCHANGED } from './constants';
import {
  defaultDeprecatedHandler,
  defaultDescriptor,
  defaultInvalidHandler,
  defaultUnknownHandler,
} from './defaults';
import { Schema } from './schema';
import {
  DeprecatedHandler,
  Descriptor,
  IdentifyMissing,
  IdentifyRequired,
  InvalidHandler,
  Logger,
  NormalizedInvalidHandler,
  NormalizedTransferResult,
  OptionKey,
  OptionPair,
  Options,
  OptionValue,
  Postprocess,
  Preprocess,
  TransferTo,
  UnknownHandler,
  Utils,
} from './types';
import {
  createAutoChecklist,
  normalizeDefaultResult,
  normalizeDeprecatedResult,
  normalizeExpectedResult,
  normalizeForwardResult,
  normalizeInvalidHandler,
  normalizeRedirectResult,
  normalizeValidateResult,
  partition,
  recordFromArray,
} from './utils';

export interface NormalizerOptions {
  logger?: false | Logger;
  loggerPrintWidth?: number;
  descriptor?: Descriptor;
  unknown?: UnknownHandler;
  invalid?: InvalidHandler;
  deprecated?: DeprecatedHandler;
  missing?: IdentifyMissing;
  required?: IdentifyRequired;
  preprocess?: Preprocess;
  postprocess?: Postprocess;
}

export const normalize = (
  options: Options,
  schemas: Array<Schema<any>>,
  opts?: NormalizerOptions,
) => new Normalizer(schemas, opts).normalize(options);

export class Normalizer {
  private _utils: Utils;
  private _unknownHandler: UnknownHandler;
  private _invalidHandler: NormalizedInvalidHandler;
  private _deprecatedHandler: DeprecatedHandler;
  private _hasDeprecationWarned!: ReturnType<typeof createAutoChecklist>;
  private _identifyMissing: IdentifyMissing;
  private _identifyRequired: IdentifyRequired;
  private _preprocess: Preprocess;
  private _postprocess: Postprocess;

  constructor(schemas: Array<Schema<any>>, opts?: NormalizerOptions) {
    // istanbul ignore next
    const {
      logger = console,
      loggerPrintWidth = 80,
      descriptor = defaultDescriptor,
      unknown = defaultUnknownHandler,
      invalid = defaultInvalidHandler,
      deprecated = defaultDeprecatedHandler,
      missing = () => false,
      required = () => false,
      preprocess = (x: Options) => x,
      postprocess = (): typeof VALUE_UNCHANGED => VALUE_UNCHANGED,
    } = opts || {};

    this._utils = {
      descriptor,
      logger: /* istanbul ignore next */ logger || { warn: () => {} },
      loggerPrintWidth,
      schemas: recordFromArray(schemas, 'name'),
      normalizeDefaultResult,
      normalizeExpectedResult,
      normalizeDeprecatedResult,
      normalizeForwardResult,
      normalizeRedirectResult,
      normalizeValidateResult,
    };

    this._unknownHandler = unknown;
    this._invalidHandler = normalizeInvalidHandler(invalid);
    this._deprecatedHandler = deprecated;
    this._identifyMissing = (k, o) => !(k in o) || missing(k, o);
    this._identifyRequired = required;
    this._preprocess = preprocess;
    this._postprocess = postprocess;

    this.cleanHistory();
  }

  public cleanHistory() {
    this._hasDeprecationWarned = createAutoChecklist();
  }

  public normalize(options: Options): Options {
    const newOptions: Options = {};

    const preprocessed = this._preprocess(options, this._utils);
    const restOptionsArray = [preprocessed];

    const applyNormalization = () => {
      while (restOptionsArray.length !== 0) {
        const currentOptions = restOptionsArray.shift()!;
        const transferredOptionsArray = this._applyNormalization(
          currentOptions,
          newOptions,
        );
        restOptionsArray.push(...transferredOptionsArray);
      }
    };

    applyNormalization();

    for (const key of Object.keys(this._utils.schemas)) {
      const schema = this._utils.schemas[key];
      if (!(key in newOptions)) {
        const defaultResult = normalizeDefaultResult(
          schema.default(this._utils),
        );
        if ('value' in defaultResult) {
          restOptionsArray.push({ [key]: defaultResult.value! });
        }
      }
    }

    applyNormalization();

    for (const key of Object.keys(this._utils.schemas)) {
      if (!(key in newOptions)) {
        continue;
      }

      const schema = this._utils.schemas[key];

      const value = newOptions[key];
      const newValue = schema.postprocess(value, this._utils);

      if (newValue === VALUE_UNCHANGED) {
        continue;
      }

      this._applyValidation(newValue, key, schema);

      newOptions[key] = newValue;
    }

    this._applyPostprocess(newOptions);
    this._applyRequiredCheck(newOptions);

    return newOptions;
  }

  private _applyNormalization(
    options: Options,
    newOptions: Options,
  ): Options[] {
    const transferredOptionsArray: Options[] = [];

    const { knownKeys, unknownKeys } = this._partitionOptionKeys(options);

    for (const key of knownKeys) {
      const schema = this._utils.schemas[key];
      const value = schema.preprocess(options[key], this._utils);

      this._applyValidation(value, key, schema);

      const appendTransferredOptions = <$Value>(
        { from, to }: NormalizedTransferResult<$Value>, //
      ) => {
        transferredOptionsArray.push(
          typeof to === 'string' ? { [to]: from } : { [to.key]: to.value },
        );
      };

      const warnDeprecated = <$Value>({
        value: currentValue,
        redirectTo,
      }: {
        value: $Value;
        redirectTo?: TransferTo;
      }) => {
        const deprecatedResult = normalizeDeprecatedResult(
          schema.deprecated(currentValue, this._utils),
          value,
          /* doNotNormalizeTrue */ true,
        );

        if (deprecatedResult === false) {
          return;
        }

        if (deprecatedResult === true) {
          if (!this._hasDeprecationWarned(key)) {
            this._utils.logger.warn(
              this._deprecatedHandler(key, redirectTo, this._utils),
            );
          }
        } else {
          for (const { value: deprecatedValue } of deprecatedResult) {
            const pair: OptionPair = { key, value: deprecatedValue };
            if (!this._hasDeprecationWarned(pair)) {
              const redirectToPair =
                typeof redirectTo === 'string'
                  ? { key: redirectTo, value: deprecatedValue }
                  : redirectTo;
              this._utils.logger.warn(
                this._deprecatedHandler(pair, redirectToPair, this._utils),
              );
            }
          }
        }
      };

      const forwardResult = normalizeForwardResult(
        schema.forward(value, this._utils),
        value,
      );

      forwardResult.forEach(appendTransferredOptions);

      const redirectResult = normalizeRedirectResult(
        schema.redirect(value, this._utils),
        value,
      );

      redirectResult.redirect.forEach(appendTransferredOptions);

      if ('remain' in redirectResult) {
        const remainingValue = redirectResult.remain!;

        newOptions[key] =
          key in newOptions
            ? schema.overlap(newOptions[key], remainingValue, this._utils)
            : remainingValue;

        warnDeprecated({ value: remainingValue });
      }

      for (const { from, to } of redirectResult.redirect) {
        warnDeprecated({ value: from, redirectTo: to });
      }
    }

    for (const key of unknownKeys) {
      const value = options[key];
      this._applyUnknownHandler(
        key,
        value,
        newOptions,
        (knownResultKey, knownResultValue) => {
          transferredOptionsArray.push({ [knownResultKey]: knownResultValue });
        },
      );
    }

    return transferredOptionsArray;
  }

  private _applyRequiredCheck(options: Options) {
    for (const key of Object.keys(this._utils.schemas)) {
      if (this._identifyMissing(key, options)) {
        if (this._identifyRequired(key)) {
          throw this._invalidHandler(key, VALUE_NOT_EXIST, this._utils);
        }
      }
    }
  }

  private _partitionOptionKeys(options: Options) {
    const [knownKeys, unknownKeys] = partition(
      Object.keys(options).filter(key => !this._identifyMissing(key, options)),
      key => key in this._utils.schemas,
    );
    return { knownKeys, unknownKeys };
  }

  private _applyValidation(value: unknown, key: string, schema: Schema<any>) {
    const validateResult = normalizeValidateResult(
      schema.validate(value, this._utils),
      value,
    );
    if (validateResult !== true) {
      throw this._invalidHandler(key, validateResult.value, this._utils);
    }
  }

  private _applyUnknownHandler(
    key: OptionKey,
    value: OptionValue,
    newOptions: Options,
    knownResultHandler: (key: OptionKey, value: OptionValue) => void,
  ) {
    const unknownResult = this._unknownHandler(key, value, this._utils);

    if (!unknownResult) {
      return;
    }

    for (const resultKey of Object.keys(unknownResult)) {
      if (this._identifyMissing(resultKey, unknownResult)) {
        continue;
      }

      const resultValue = unknownResult[resultKey];
      if (resultKey in this._utils.schemas) {
        knownResultHandler(resultKey, resultValue);
      } else {
        newOptions[resultKey] = resultValue;
      }
    }
  }

  private _applyPostprocess(options: Options): void {
    const postprocessed = this._postprocess(options, this._utils);

    if (postprocessed === VALUE_UNCHANGED) {
      return;
    }

    if (postprocessed.delete) {
      for (const deleteKey of postprocessed.delete) {
        delete options[deleteKey];
      }
    }

    if (postprocessed.override) {
      const { knownKeys, unknownKeys } = this._partitionOptionKeys(
        postprocessed.override,
      );

      for (const key of knownKeys) {
        const value = postprocessed.override[key];
        this._applyValidation(value, key, this._utils.schemas[key]);
        options[key] = value;
      }

      for (const key of unknownKeys) {
        const value = postprocessed.override[key];
        this._applyUnknownHandler(
          key,
          value,
          options,
          (knownResultKey, knownResultValue) => {
            const schema = this._utils.schemas[knownResultKey];
            this._applyValidation(knownResultValue, knownResultKey, schema);
            options[knownResultKey] = knownResultValue;
          },
        );
      }
    }
  }
}
