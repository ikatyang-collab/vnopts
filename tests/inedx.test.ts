import * as vnopts from '../src';
import { IdentifyMissing } from '../src';
import { createLogger } from './__helpers__/utils';

const logger = createLogger();

beforeEach(() => {
  logger.clearMessages();
});

const schemas = [
  vnopts.createSchema(vnopts.BooleanSchema, {
    name: 'useFlowParser',
    deprecated: true,
    redirect: (value: boolean) =>
      !value ? undefined : { to: { key: 'parser', value: 'flow' } },
  }),
  vnopts.createSchema(vnopts.ChoiceSchema, {
    name: 'parser',
    validate: (value: unknown) =>
      ['function', 'string'].indexOf(typeof value) !== -1,
    choices: [
      'flow',
      'babylon',
      'typescript',
      {
        value: 'postcss',
        deprecated: true,
        redirect: { key: 'parser', value: 'css' },
      },
      'css',
      'less',
      'scss',
    ],
  }),
];

test('deprecated', () => {
  expect(
    vnopts.normalize({ useFlowParser: true }, schemas, { logger }),
  ).toMatchSnapshot();
  expect(logger.getMessages()).toMatchSnapshot();
});

test('redirect', () => {
  expect(
    vnopts.normalize({ parser: 'postcss' }, schemas, { logger }),
  ).toMatchSnapshot();
  expect(logger.getMessages()).toMatchSnapshot();
});

describe('missing', () => {
  const name = 'a';
  const missing: IdentifyMissing = (key, options) => options[key] === undefined;

  test('missing pair will be filtered', () => {
    expect(
      vnopts.normalize(
        { [name]: undefined },
        [vnopts.createSchema(vnopts.AnySchema, { name, validate: false })],
        { missing },
      ),
    ).toEqual({});
  });

  test('missing pair will be replaced by default pair if present', () => {
    const defaultValue = 'foo';
    expect(
      vnopts.normalize(
        { [name]: undefined },
        [
          vnopts.createSchema(vnopts.AnySchema, {
            name,
            default: { value: defaultValue },
          }),
        ],
        { missing },
      ),
    ).toEqual({ a: defaultValue });
  });
});

test('required', () => {
  expect(() =>
    vnopts.normalize(
      {},
      [vnopts.createSchema(vnopts.AnySchema, { name: '<key>' })],
      { required: () => true },
    ),
  ).toThrowErrorMatchingSnapshot();
});
