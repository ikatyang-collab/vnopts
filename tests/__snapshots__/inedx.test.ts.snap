// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deprecated 1`] = `
Object {
  "parser": "flow",
}
`;

exports[`deprecated 2`] = `
Array [
  "useFlowParser is deprecated; we now treat it as { parser: \\"flow\\" }.",
]
`;

exports[`postprocess schema throw error for invalid postprocessed value 1`] = `"Invalid \\"<key>\\" value. Expected \\"<valid-value-1>\\", but received \\"<invalid-value>\\"."`;

exports[`redirect 1`] = `
Object {
  "parser": "css",
}
`;

exports[`redirect 2`] = `
Array [
  "{ parser: \\"postcss\\" } is deprecated; we now treat it as { parser: \\"css\\" }.",
]
`;

exports[`required 1`] = `"Invalid \\"<key>\\" value. Expected anything, but received nothing."`;
